accord-core/src/main/java/accord/local/CommandStores.java

```diff
@@ -356,42 +315,110 @@ public synchronized void shutdown()
+        return reduce(futures, reduce);
+    }
+
+    private  <S, T> T mapReduce(TxnOperation operation, Select<S> select, S scope, long minEpoch, long maxEpoch, Fold<TxnOperation, Void, List<Future<T>>> fold, BiFunction<T, T, T> reduce)
+    {
+        List<Future<T>> futures = foldl(select, scope, minEpoch, maxEpoch, fold, operation, null, ArrayList::new);
+        if (futures == null)
+            return null;
+        return reduce(futures, reduce);
+    }
+
+    public <T> T mapReduce(TxnOperation operation, Key key, long minEpoch, long maxEpoch, Function<CommandStore, T> map, BiFunction<T, T, T> reduce)
+    {
+        return mapReduce(operation, ShardedRanges::shard, key, minEpoch, maxEpoch, mapReduceFold(map), reduce);
+    }
+
+    public <T> T mapReduce(TxnOperation operation, Key key, long epoch, Function<CommandStore, T> map, BiFunction<T, T, T> reduce)
+    {
+        return mapReduce(operation, key, epoch, epoch, map, reduce);
+    }
+
+    public <T> T mapReduceSince(TxnOperation operation, Key key, long epoch, Function<CommandStore, T> map, BiFunction<T, T, T> reduce)
+    {
+        return mapReduce(operation, key, epoch, Long.MAX_VALUE, map, reduce);
+    }
 
-    public void forEach(Consumer<CommandStore> forEach)
+    public <T> T mapReduce(TxnOperation operation, Keys keys, long minEpoch, long maxEpoch, Function<CommandStore, T> map, BiFunction<T, T, T> reduce)
     {
-        forEach((s, i, min, max) -> s.all(), null, 0, 0, forEach);
+        // probably need to split txnOperation and scope stuff here
+        return mapReduce(operation, ShardedRanges::shards, keys, minEpoch, maxEpoch, mapReduceFold(map), reduce);
     }
 
-    public void forEach(Keys keys, long epoch, Consumer<CommandStore> forEach)
+    public void setup(Consumer<CommandStore> forEach)
     {
-        forEach(keys, epoch, epoch, forEach);
+        setup(forEach, (store, f, i, t) -> { t.add(store.processSetup(f)); return t; }, (Void i1, Void i2) -> null);
     }
 
-    public void forEach(Keys keys, long minEpoch, long maxEpoch, Consumer<CommandStore> forEach)
+    public <T> T setup(Function<CommandStore, T> map, BiFunction<T, T, T> reduce)
     {
-        forEach(ShardedRanges::shards, keys, minEpoch, maxEpoch, forEach);
+        return setup(map, (store, f, i, t) -> { t.add(store.processSetup(f)); return t; }, reduce);
     }
 
-    public <T> T mapReduce(Keys keys, long epoch, Function<CommandStore, T> map, BiFunction<T, T, T> reduce)
+    private static Fold<TxnOperation, Void, List<Future<Void>>> forEachFold(Consumer<CommandStore> forEach)
     {
-        return mapReduce(keys, epoch, epoch, map, reduce);
+        return (store, op, i, t) -> { t.add(store.process(op, forEach)); return t; };
     }
 
-    public <T> T mapReduce(Keys keys, long minEpoch, long maxEpoch, Function<CommandStore, T> map, BiFunction<T, T, T> reduce)
+    public void forEach(TxnOperation operation, Keys keys, long minEpoch, long maxEpoch, Consumer<CommandStore> forEach)
```

Comment:
"forEach" takes both a "TxnOperation" and "Keys", but "TxnOperation" holds ref to "Keys", so shouldn't we drop "Keys"? Looking at the usage, it seems like we could...

```
node.forEachLocalSince(TxnOperation.scopeFor(txnId), full.txn.keys(), ...
```

```
TxnOperation scope = TxnOperation.scopeFor(txnId, max.txn.keys());
...
node.forEachLocalSince(scope, max.txn.keys(), ...
```

This change trickled throughout a lot of different places, so would be good to grasp the expected semantics... if the TxnOperation has keys, and you provide keys... "should" they be released or "must" they be related? or something else?

Reply:
Reply from Blake in JIRA

The keys in each perform different functions. The keys in the TxnOperation indicate which CommandsForKey we need access to, this let's C* know which ones it needs to load. The Keys passed into forEach determine which command stores we need to talk to, since we need to talk to "replicas" of the given keys, whether we need the commands for key to be in memory or not. Of course, this isn't obvious without more complete comments on TxnOperation

> This change trickled throughout a lot of different places, so would be good to grasp the expected semantics... if the TxnOperation has keys, and you provide keys... "should" they be released or "must" they be related? or something else?

I added some comments. The name TxnOperation is a little unclear. Let me know if you have any suggestions for better names. Maybe TxnOperationScope? Maybe it could extend Function<CommandStore, T> and bundle the operation and the metadata into the same interface? Calls would be a bit more cumbersome, but would make the interdependency of the 3 parts more obvious

Reply:
I think we should rename `TxnOperation` to something like `PreLoadContext` to indicate its purpose, which is to ensure the necessary state is ready for the operation before it executes.